### ISSUE TITLE
BooleanMap/Set implementations should delegate to a super simple data structure

### ISSUE THREAD

**Issue by techsy730** [vigna/fastutil#131]
Instead of having grand data structures for a keyspace of only 2 elements, all BooleanMap/Set implementations should delegate to a super simple implementation that only has whether `true` is present, whether `false` is present, and for maps, the `true` value and the `false` value. This should be super fast and super compact in memory.

The only complexity would be that for some implementations, a sorted order needs to be maintained (those that implement SortedBooleanMap/Set). In others, we need to maintain insertion order. (For implementations that do not assure a key order, we can just use the sorted order for free).

**Comment by techsy730** [comment]
Oh, I didn't notice that there aren't even Boolean2WhateverMaps generated. So I guess make this bug only about BooleanSet

**Comment by vigna** [comment]
"All BooleanMap/Set implementations" sounds a bit _grand_ for a single implementation: BooleanOpenHashSet. LOL.

**Comment by techsy730** [comment]
Yep, the whole point is that something like "BooleanOpenHashSet" is complete overkill for booleans. It should just be two boolean variables (whether true and false are present). All of the boolean sets should be. The only reason I didn't suggest getting rid of all the boolean set implementations and just have one or two of this sort of super simple implementation is for backwards compatibility.

**Comment by vigna** [comment]
Ok. In the interest of clarity, can you name another implementation beside BooleanOpenHashSet?

**Comment by techsy730** [comment]
Hmm, looking over it, BooleanOpenHashSet is the only concrete set type for booleans currently. There isn't even a SortedSet implementation.

So I guess then we can just make a new set implementation, BooleanSimpleSet or BooleanFastSet or BooleanDirectSet or something, that is just two boolean member variables.

Optionally, we can then make BooleanOpenHashSet delegate to that implementation.

**Comment by leventov** [comment]
+1. BooleanOpenHashSet and BooleanArraySet should be deprecated.

**Comment by vigna** [comment]
I'd just write ten-line specific implementations of those, with a two-element boolean array remembering whether true or false are in the set. Pull requests, anybody?

**Comment by techsy730** [comment]
I might be willing to try this. I just had my hands full with the Spliterator stuff at the time.
However, implementing NavigableMap may be a better first step.

**Comment by vigna** [comment]
Yes I would try to release  8.5.0 with #162 and the attack navigabile. There's even a proposal from _2017_ with the interfaces 🙈.

**Comment by techsy730** [comment]
Given that a boolean _set_ can have only 4 possible states, {}, {false}, {true}, and {false, true}, it would make sense to have an immutable boolean set implementation alongside the mutable one.
